File: src/main/java/dev/kikugie/techutils/feature/GiveFullIInv.java
package dev.kikugie.techutils.feature;

import dev.kikugie.techutils.TechUtilsMod;
import dev.kikugie.techutils.config.MiscConfigs;
import fi.dy.masa.malilib.util.game.BlockUtils;
import net.minecraft.block.Block;
import net.minecraft.block.BlockEntityProvider;
import net.minecraft.block.Blocks;
import net.minecraft.block.ShulkerBoxBlock;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.ChestBlockEntity;
import net.minecraft.block.entity.LootableContainerBlockEntity;
import net.minecraft.block.entity.ShulkerBoxBlockEntity;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.network.ClientPlayerEntity;
import net.minecraft.component.DataComponentTypes;
import net.minecraft.component.type.BundleContentsComponent;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.*;
import net.minecraft.text.Text;
import net.minecraft.util.DyeColor;
import net.minecraft.util.Formatting;
import net.minecraft.util.math.BlockPos;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 * Inserts a full container of given item into player's hand. Works <b>only</b> in creative mode.
 */
public class GiveFullIInv {
	private static final GiveFullIInv INSTANCE = new GiveFullIInv();
	private static final Supplier<Boolean> SAFETY = MiscConfigs.FILL_SAFETY::getBooleanValue;

	public static boolean onKeybind() {
		ClientPlayerEntity player = MinecraftClient.getInstance().player;
		assert player != null;
		if (!player.isCreative()) {
			INSTANCE.sendError("not_creative_enough");
			return false;
		}

		ItemStack mainHand = player.getMainHandStack();
		ItemStack offHand = player.getOffHandStack();

		Optional<ItemStack> result = get(mainHand, offHand);
		if (result.isEmpty())
			return false;
		int selectedSlot = player.getInventory().getSelectedSlot();
		player.getInventory().setStack(selectedSlot, result.get());
		Objects.requireNonNull(MinecraftClient.getInstance().interactionManager).clickCreativeStack(result.get(), 36 + selectedSlot);
		player.playerScreenHandler.sendContentUpdates();
		return true;
	}

	public static Optional<ItemStack> get(ItemStack mainHand, ItemStack offHand) {
		return INSTANCE.getItem(mainHand, offHand);
	}

	public static ItemStack fillShulker(ItemStack stack, @Nullable DyeColor color) {
		Block shulker = ShulkerBoxBlock.get(color);
		ShulkerBoxBlockEntity box = new ShulkerBoxBlockEntity(BlockPos.ORIGIN, shulker.getDefaultState());
		return fillLootable(stack, shulker.asItem(), box);
	}

	public static ItemStack fillChest(ItemStack stack) {
		Block chest = Blocks.CHEST;
		ChestBlockEntity box = new ChestBlockEntity(BlockPos.ORIGIN, chest.getDefaultState());
		return fillLootable(stack, chest.asItem(), box);
	}

	public static ItemStack fillLootable(ItemStack stack, Item item, LootableContainerBlockEntity lootable) {
		for (int i = 0; i < lootable.size(); i++) {
			lootable.setStack(i, stack);
		}
		ItemStack container = item.getDefaultStack();
		BlockUtils.setStackNbt(container, lootable, MinecraftClient.getInstance().world.getRegistryManager());
		return container;
	}

	public static ItemStack fillBundle(ItemStack stack) {
		List<ItemStack> stacks = new ArrayList<>();
		for (int i = 0; i < MiscConfigs.BUNDLE_FILL.getIntegerValue(); i++) {
			stacks.add(stack.copy());
		}
		ItemStack bundle = Items.BUNDLE.getDefaultStack();
		BundleContentsComponent.Builder builder = new BundleContentsComponent.Builder(new BundleContentsComponent(stacks));
		bundle.set(DataComponentTypes.BUNDLE_CONTENTS, builder.build());
		return bundle;
	}

	@SuppressWarnings("DataFlowIssue")
	public static boolean containerHasItems(ItemStack container) {
		if (container.getItem() instanceof BlockItem blockItem && blockItem.getBlock() instanceof BlockEntityProvider provider) {
			var container_data = container.get(DataComponentTypes.CONTAINER);
			if (container_data == null)
				return false;

			BlockEntity blockEntity = provider.createBlockEntity(BlockPos.ORIGIN, blockItem.getBlock().getDefaultState());
			blockEntity.readComponents(container);
			if (blockEntity instanceof Inventory inventory)
				return !inventory.isEmpty();
		}
		return false;
	}

	public static boolean bundleHasItems(ItemStack bundle) {
		return BundleItem.getAmountFilled(bundle) > 0;
	}

	private static String generateCommand(ItemStack stack, int slot) {
		String template = "item replace entity @s container.%s with %s 1";
		return "";
	}

	private Optional<ItemStack> getItem(ItemStack mainHand, ItemStack offHand) {
		if (mainHand.isEmpty()) {
			sendError("no_item");
			return Optional.empty();
		}

		return isShulkerBox(mainHand) ? handleBox(mainHand, offHand) : handleItem(mainHand, offHand);
	}

	private Optional<ItemStack> handleItem(ItemStack mainHand, ItemStack offHand) {
		if (!SAFETY.get() && !recursionCheck(mainHand)) {
			sendError("nested_stack");
			return Optional.empty();
		}
		ItemStack fullStack = mainHand.copyWithCount(mainHand.getMaxCount());
		return Optional.of(handleOffHand(offHand, stack -> fillShulker(stack, null)).apply(fullStack));
	}

	private Optional<ItemStack> handleBox(ItemStack mainHand, ItemStack offHand) {
		if (!SAFETY.get() && isShulkerBox(offHand)) {
			sendError("nested_box");
			return Optional.empty();
		}
		ItemStack fullStack = containerHasItems(mainHand) ? mainHand.copy() : mainHand.copyWithCount(64);
		return Optional.of(handleOffHand(offHand, GiveFullIInv::fillChest).apply(fullStack));
	}

	private boolean recursionCheck(ItemStack mainHand) {
		if (mainHand.getItem() instanceof BundleItem)
			return !bundleHasItems(mainHand);
		return !containerHasItems(mainHand);
	}

	private Function<ItemStack, ItemStack> handleOffHand(ItemStack offHand, @Nullable Function<ItemStack, ItemStack> fallback) {
		if (offHand.isEmpty())
			return fallback;
		// Item has a corresponding block entity
		if (offHand.getItem() instanceof BlockItem blockItem && blockItem.getBlock() instanceof BlockEntityProvider provider) {
			BlockEntity blockEntity = provider.createBlockEntity(BlockPos.ORIGIN, blockItem.getBlock().getDefaultState());
			// Block entity is a container
			if (blockEntity instanceof LootableContainerBlockEntity lootable)
				return stack -> fillLootable(stack, blockItem, lootable);
		}
		if (offHand.getItem() instanceof BundleItem) {
			return GiveFullIInv::fillBundle;
		}
		// Some other item
		return fallback;
	}

	private void sendError(String key) {
		Text message = Text.translatable("techutils.feature.givefullinv." + key).formatted(Formatting.DARK_RED);
		if (MinecraftClient.getInstance() != null && MinecraftClient.getInstance().player != null)
			MinecraftClient.getInstance().player.sendMessage(message, true);
		else
			TechUtilsMod.LOGGER.warn(message.getString());
	}

	private boolean isShulkerBox(ItemStack stack) {
		return stack.getItem().toString().contains("shulker_box");
	}
}

File: src/main/kotlin/dev/kikugie/techutils/client/feature/giveinv/GiveFullIInv.kt
package dev.kikugie.techutils.client.feature.giveinv

import net.minecraft.block.BlockEntityProvider
import net.minecraft.block.Blocks
import net.minecraft.block.ShulkerBoxBlock
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.ChestBlockEntity
import net.minecraft.block.entity.LootableContainerBlockEntity
import net.minecraft.block.entity.ShulkerBoxBlockEntity
import net.minecraft.client.MinecraftClient
import net.minecraft.item.*
import net.minecraft.nbt.NbtCompound
import net.minecraft.nbt.NbtElement
import net.minecraft.nbt.NbtList
import net.minecraft.util.DyeColor
import net.minecraft.util.math.BlockPos
import java.util.function.Function

/**
 * Inserts a full container of given item into player's hand. Works **only** in creative mode.
 */
object GiveFullIInv {
    private val safety = { GiveFullInvConfig.fillSafety.booleanValue }
    private val bundleFill = { GiveFullInvConfig.bundleFill.integerValue }
    fun onKeybind(): Boolean {
        val client = MinecraftClient.getInstance()
        val player = client.player!!
        if (!player.isCreative) {
            sendError("not_creative_enough")
            return false
        }
        val mainHand = player.mainHandStack.copy()
        val offHand = player.offHandStack.copy()

        val result: ItemStack = getStack(mainHand, offHand) ?: return false
        client.interactionManager?.clickCreativeStack(result, 36 + player.inventory.selectedSlot)
        return true
    }

    fun getStack(mainHand: ItemStack, offHand: ItemStack): ItemStack? {
        if (mainHand.isEmpty) {
            sendError("no_item")
            return null
        }

        return if (isShulkerBox(mainHand))
            handleBox(mainHand, offHand)
        else handleItem(mainHand, offHand)
    }

    private fun handleBox(mainHand: ItemStack, offHand: ItemStack): ItemStack? {
        if (safety.invoke() && isShulkerBox(offHand)) {
            sendError("nested_box")
            return null
        }
        val fullStack = if (containerHasItems(mainHand))
            mainHand.copy()
        else
            mainHand.copyWithCount(64)
        return handleOffHand(offHand) { stack: ItemStack -> fillChest(stack) }.apply(fullStack)
    }

    private fun handleItem(mainHand: ItemStack, offHand: ItemStack): ItemStack? {
        if (safety.invoke() && !checkRecursion(mainHand)) {
            sendError("nested_stack")
            return null
        }
        val fullStack = mainHand.copyWithCount(mainHand.maxCount)
        return handleOffHand(offHand) { stack: ItemStack ->
            fillShulker(stack, null)
        }.apply(fullStack)
    }

    private fun handleOffHand(
        offHand: ItemStack,
        fallback: Function<ItemStack, ItemStack>
    ): Function<ItemStack, ItemStack> {
        if (offHand.isEmpty) return fallback
        // Item has a corresponding block entity
        if (offHand.item is BlockItem && (offHand.item as BlockItem).block is BlockEntityProvider) {
            val blockItem = offHand.item as BlockItem
            val provider = blockItem.block as BlockEntityProvider
            val blockEntity: BlockEntity? = provider.createBlockEntity(BlockPos.ORIGIN, blockItem.block.defaultState)
            // Block entity is a container
            if (blockEntity is LootableContainerBlockEntity)
                return Function<ItemStack, ItemStack> { stack: ItemStack ->
                    fillLootable(
                        stack,
                        blockItem,
                        blockEntity
                    )
                }
        }
        return if (offHand.item is BundleItem) {
            Function<ItemStack, ItemStack> { stack: ItemStack -> fillBundle(stack) }
        } else fallback
    }

    private fun fillShulker(stack: ItemStack, color: DyeColor?): ItemStack {
        val shulker = ShulkerBoxBlock.get(color)
        val box = ShulkerBoxBlockEntity(BlockPos.ORIGIN, shulker.defaultState)
        return fillLootable(stack, shulker.asItem(), box)
    }

    private fun fillChest(stack: ItemStack): ItemStack {
        val chest = Blocks.CHEST
        val box = ChestBlockEntity(BlockPos.ORIGIN, chest.defaultState)
        return fillLootable(stack, chest.asItem(), box)
    }

    private fun fillLootable(stack: ItemStack, item: Item, lootable: LootableContainerBlockEntity): ItemStack {
        for (i in 0 until lootable.size()) {
            lootable.setStack(i, stack)
        }
        val container = item.defaultStack
        lootable.setStackNbt(container)
        return container
    }

    private fun fillBundle(stack: ItemStack): ItemStack {
        val bundle = Items.BUNDLE.defaultStack
        val nbtCompound = bundle.getOrCreateNbt()
        if (!nbtCompound.contains("Items")) {
            nbtCompound.put("Items", NbtList())
        }
        val nbtList = nbtCompound.getList("Items", NbtElement.COMPOUND_TYPE.toInt())
        val nbt = NbtCompound()
        stack.writeNbt(nbt)
        for (i in 0 until bundleFill.invoke()) {
            nbtList.add(nbt.copy())
        }
        return bundle
    }

    private fun checkRecursion(mainHand: ItemStack): Boolean {
        if (!mainHand.hasNbt()) return true
        return if (mainHand.item is BundleItem)
            !bundleHasItems(mainHand)
        else !containerHasItems(mainHand)
    }

    private fun isShulkerBox(stack: ItemStack): Boolean {
        return stack.item.toString().contains("shulker_box")
    }

    private fun containerHasItems(container: ItemStack): Boolean {
        if (container.item !is BlockItem || (container.item as BlockItem).block !is BlockEntityProvider)
            return false
        return container.getOrCreateSubNbt("BlockEntityTag").contains("Items")
    }

    private fun bundleHasItems(bundle: ItemStack): Boolean {
        val nbtCompound = bundle.getOrCreateNbt()
        return if (!nbtCompound.contains("Items"))
            false
        else !nbtCompound.getList(
            "Items",
            NbtElement.COMPOUND_TYPE.toInt()
        ).isEmpty()
    }

    private fun sendError(error: String) {
    }
}
